dist_q_learning/q_estimators.py

```diff
@@ -584,232 +596,177 @@ def update(self, history_batch):
-                    print(f"Scaled q_ais={q_ais} fake_q={fake_q_ais}"
-                          f"ns={n_ais}")
-
-                # in the original algorithm this min was also over the
-                # quantiles as well as the fake targets
-                # TODO - I think this takes a while to ever get non-negative -
-                #  see above
-                trans_ns = np.min(n_ais, axis=-1)
-                if states.shape[0] <= 10:
-                    print(f"ns before zeroed {trans_ns}")
-                trans_ns = np.maximum(trans_ns, 0.)
-
-                # TODO - consider clip at [0, 1]?
-                outside_0_1 = np.logical_or(q_ais < 0., q_ais > 1.)
-                if np.any(outside_0_1) and self.scaled:
-                    print(f"WARN: some Q means outside 0, 1: "
-                          f"{outside_0_1.sum()}/{q_ais.shape[0]} values")
-                    q_ais = np.minimum(np.maximum(q_ais, 0.), 1.)
-
-                q_alphas = q_ais * trans_ns + 1.
-                q_betas = (1. - q_ais) * trans_ns + 1.
-                if states.shape[0] <= 10:
-                    print(f"q_alpha={q_alphas}, q_beta={q_betas}")
-
-                q_target_transitions = np.squeeze(
-                    scipy.stats.beta.ppf(self.quantile, q_alphas, q_betas), 0)
-                assert q_target_transitions.shape[0] == idxs.shape[0], (
-                    f"{q_target_transitions.shape}, {idxs.shape}")
-
+                print(f"Future Q {future_qs}")
+
+            ire_ns, ire_alphas, ire_betas = ire.model.uncertainty_estimate(
+                states=states,
+                x_batch=conv_states,
+                y_batch=conv_rewards,
+                debug=debug)
+            IV_is = scipy.stats.beta.ppf(
+                self.quantile, ire_alphas, ire_betas)
+            if debug:
+                print(f"s=\n{states}")
+                print(f"IRE alphas=\n{ire_alphas}\nbetas=\n{ire_betas}")
+                print(f"IV_is at q_({self.quantile:.4f}):\n{IV_is}")
+
+            # Q target = r + (h-1)-step future from next state (future_qs)
+            # so to scale q to (0, 1) (exc gamma), scale by (~h)
+            if self.horizon_type == "inf":
+                scaled_IV_is = (
+                    1. - self.gamma) * IV_is if self.scaled else IV_is
+                q_targets = scaled_IV_is + self.gamma * future_qs
+            else:
+                # TODO - incorporate gamma or keep approx?
+                q_targets = IV_is / h + future_qs * (h - 1) / h
+            if debug:
+                print("Doing update using IRE uncertainty...")
+                print(f"s=\n{states}")
+                print(f"action=\n{update_action}")
+                print(f"IRE q_targets combined=\n{q_targets}")
+            # TODO lr must be scalar... Also what?
+            # TODO - do this update at the end, rather than use the model
+            #  for the next transition estimate?
```

Comment:
I think it's correct to do the update now, then do the transition update...? At least thats what we did for the grid world, and is what is written in the fabled Algorithm 3